l10n_nl_postcodeapi/models/res_partner.py

```diff
@@ -0,0 +1,111 @@
+#
+#    @autors: Stefan Rijnhart, Ronald Portier
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from lxml import etree
+from openerp import models, api, exceptions, _
+from openerp.tools import ormcache
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    @ormcache(skiparg=2)
+    def get_provider_obj(self):
+        self._cr.execute("""
+                SELECT value FROM ir_config_parameter
+                WHERE key = %s""", ('l10n_nl_postcodeapi.apikey',))
+        row = self._cr.fetchone()
+        if row and row[0] != 'Your API key' and row[0].strip():
+            from pyPostcode import Api
+            provider = Api(row[0].strip())
+            test = provider.getaddress('1053NJ', '334T')
+            if not test or not test._data:
+                raise exceptions.Warning(
+                    _('Error'),
+                    _('Could not verify the connection with the address '
+                      'lookup service (if you want to get rid of this '
+                      'message, please rename or delete the system parameter '
+                      '\'l10n_nl_postcodeapi.apikey\').'))
+            return provider
+        return False
+
+    @ormcache(skiparg=1)
+    def get_province(self, province):
+        """ Return the province or empty recordset """
+        if not province:
+            return self.env['res.country.state']
+        res = self.env['res.country.state'].search([('name', '=', province)])
+        return res[0] if res else res
+
+    @api.onchange('zip', 'street_number', 'country_id')
+    def on_change_zip_street_number(self):
+        """
+        Normalize the zip code, check on the partner's country and
+        if all is well, request address autocompletion data.
+
+        NB. postal_code is named 'zip' in OpenERP, but is this a reserved
+        keyword in Python
+        """
+        postal_code = self.zip and self.zip.replace(' ', '')
+        if (not (postal_code and self.street_number)
+                or (self.country_id and
+                    self.country_id != self.env.ref('base.nl'))):
+            return {}
+
+        provider_obj = self.get_provider_obj()
+        if not provider_obj:
```

Comment:
Then you get an exception every time that you enter an address, instead of when you change the API key.